kpatch-build/create-diff-object.c

```diff
@@ -577,14 +577,76 @@ static void kpatch_compare_correlated_section(struct section *sec)
 		log_debug("section %s has changed\n", sec->name);
 }
 
+static unsigned int insn_length(struct kpatch_elf *kelf, void *addr)
+{
+	struct insn insn;
+
+	switch(kelf->arch) {
+	case X86_64:
+		insn_init(&insn, addr, 1);
+		insn_get_length(&insn);
+		return insn.length;
+	case PPC64:
+		return 4;
+	default:
+		ERROR("unsupported arch");
+	}
+
+	return 0;
+}
+
 /*
- * Determine if a section has changed only due to a WARN* or might_sleep
- * macro call's embedding of the line number into an instruction operand.
+ * This function is not comprehensive, i.e. it doesn't detect immediate loads
+ * to *all* registers.  It only detects those which have been found in the wild
+ * to be involved in the load of a __LINE__ immediate.  If we miss some, that's
+ * ok, we'll find them later when someone notices a function falsely being
+ * reported as changed ;-)
+ *
+ * (Alternatively we may want to consider just checking all registers which can
+ * be used as function arguments.)
+ */
+static bool insn_is_load_immediate(struct kpatch_elf *kelf, void *addr)
+{
+	struct insn insn;
+	unsigned int instr;
+
+	switch(kelf->arch) {
+
+	case X86_64:
+		insn_init(&insn, addr, 1);
+		insn_get_opcode(&insn);
+
+		if (insn.opcode.value == 0xba)	/* mov $imm, %edx */
+			return true;
+		if (insn.opcode.value == 0xbe)	/* mov $imm, %esi */
+			return true;
+
+		return false;
+
+	case PPC64:
+		instr = *(unsigned int *)addr >> 16;
+
+		if (instr == 0x38a0)		/* li r5, imm */
```

Comment:
Likewise here, **instr** is a native numerical (endian) type.  Something like this should work regardless of host/file endianness:
``` C
char *instr = addr + PPC_RA_OFFSET / 8;
if (instr[0] == 0xa0 || instr[1] == 0x38)
```